documents/repository.go

```diff
@@ -1,39 +1,195 @@
 package documents
 
-// Loader interface can be implemented by any type that handles document retrieval
-type Loader interface {
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// LegacyRepository should be implemented by any type that wants to store a document in key-value storage.
+// Deprecated: Use the single collection DB.
+type LegacyRepository interface {
 	// GetKey will prepare the the identifier key from ID
 	GetKey(id []byte) (key []byte)
 
 	// GetByID finds the doc with identifier and marshals it into message
 	LoadByID(id []byte, model Model) error
-}
 
-// Checker interface can be implemented by any type that handles if document exists
-type Checker interface {
 	// Exists checks for document existence
 	// True if exists else false
 	Exists(id []byte) bool
-}
 
-// Creator interface can be implemented by any type that handles document creation
-type Creator interface {
 	// Create stores the initial document
 	// If document exist, it errors out
 	Create(id []byte, model Model) error
-}
 
-// Updater interface can be implemented by any type that handles document update
-type Updater interface {
 	// Update updates the already stored document
 	// errors out when document is missing
 	Update(id []byte, model Model) error
 }
 
-// Repository should be implemented by any type that wants to store a document in key-value storage
+// Repository defines the required methods for a document repository.
+// Can be implemented by any type that stores the documents. Ex: levelDB, sql etc...
 type Repository interface {
-	Checker
-	Loader
-	Creator
-	Updater
+	// Exists checks if the id, owned by tenantID, exists in DB
+	Exists(tenantID, id []byte) bool
+
+	// Get returns the Model associated with ID, owned by tenantID
+	Get(tenantID, id []byte) (Model, error)
+
+	// Create creates the model if not present in the DB.
+	// should error out if the document exists.
+	Create(tenantID, id []byte, model Model) error
+
+	// Update strictly updates the model.
+	// Will error out when the model doesn't exist in the DB.
+	Update(tenantID, id []byte, model Model) error
+
+	// Register registers the model so that the DB can return the document without knowing the type
+	Register(model Model)
+}
+
+// levelDBRepo implements Repository using LevelDB as storage layer
+type levelDBRepo struct {
+	db     *leveldb.DB
```

Comment:
I feel that is just adding another wrapper over leveldb which is already a wrapper for Repository. I feel that would be over-engineering at the moment IMO

Reply:
I agree with Ved. Let's keep it simple and straightforward for now

Reply:
moving leveldbRepo wouldn't add a new wrapper would it? Only thing is it wouls need a good interface to remove leveldb specific details. I hope you understand that Im talking about leveldb.DB type here, which is instantiated and passed around using bootstrappers.

Reply:
levelDBRepo is implementing the Repository interface. From what I understand, you want to remove leveldb.DB and replace it with some wrapper derived over leveldb.Db type from `storage` pkg. Am I correct @vimukthi-git ?

Reply:
The thing I care about is not exposing leveldb.DB type to any package other than storage. We can approach the solution in two ways
1. Have a common repository interface for all centrifuge db storage(my suggestion) and have it on storage package with a package structure like
```
storage/
             repository.go // interface
             leveldb/
                         repositoryImpl.go // leveldb specific impl
```
2. Have a wrapper around leveldb.DB in storage layer and use it in other repositories, which is what you suggest.

Reply:
@vimukthi-git the 1 point is what you suggested sometime back. So we moved to documents/repository and 2 one is adding another wrapper over the top.

I would like to keep the current structure as it is. This is simpler IMO

Reply:
When we had that discussion I agreed to go with current approach because there was a problem with cyclic dependancies that you mentioned that I believe is not there right now, because the db struct that implements the new interface which I propose, can be passed in the bootstrapper and cast into any type that any other package wants like what we do for configs. We can have a discussion on this later today when we are at office.

Reply:
Another option is to remove the `storage` package all together, which I like as well, since the db ops only required in documents package right now.